Make airspeed sensor configuration easier and consistent with other sensors

sfuhrer opened the pull request:
### Solved Problem
Configuring whether airspeed should be used or not is more complex than it has to be. It currently is done:
- `CBRK_AIRSPD_CHK`: disables the airspeed preflight check if set to the magic number
- `FW_ARSP_MODE`: if set to 1 airspeed is not used in the controllers (rate, attitude, position, VOTL). Additionally, if set to 1, the `CBRK_AIRSPD_CHK` is not needed anymore.
- `ASPD_PRIMARY`: if set to -1 the airspeed_validated.airspeed_calibrated/true is always published as NAN
- `EKF2_ARSP_THR`: governs usage of airspeed data in the EKF

Additionally, the airspeed setpoint initialization in the fw_pos_control is more complex than it probably has to be.

### Solution
- replace `CBRK_AIRSPD_CHK` by new param `SYS_HAS_NUM_ASPD` (governs airspeed preflight check)
  - aligns it with the similar logic for baro, mag, GPS
- rename `FW_ARSP_MODE` to `FW_USE_AIRSPD`, and do not use in airspeed preflight check
  - clearer param name
  - functionality stays the same (disables airspeed usage for output scaling (rate controller), turn coordination (attitude), airspeed setpoint tracking (position), takeoff logic (position), VTOL transition logic)
- remove "Disabled" option from `ASPD_PRIMARY`
  - this is just a redundant way of disabling airspeed. If no airspeed should be used then the driver can be switched off, or then the usage can be disabled (FW_USE_AIRSPD, EKF2_ARSP_THR)
- [code clean up /commenting] be clearer when airspeed is `true` or `calibrated`/`equivalent`
- airspeed setpoint initialization in fw_pos_control: initialize with current airspeed (limited to FW_AIRSPD_MIN) --> remove dedicated VTOL airspeed init after transition as it's not needed (there is already a slew rate limit on the setpoint, and it should just be init at the current airspeed)

### Changelog Entry
For release notes:
```
Feature: Make airspeed sensor configuration easier and consistent with other sensors.
New parameter: SYS_HAS_NUM_ASPD, FW_USE_AIRSPD
Documentation: TODO
```

### Test coverage
SITL testing.

sfuhrer commented:
@RomanBapst @dagar @tstastny could you have a look here and let me know what you think of these parameter renames/functional changes in particular? Thanks!

KonradRudin commented on src/modules/fw_pos_control/FixedwingPositionControl.cpp:
```diff
@@ -391,19 +385,16 @@ FixedwingPositionControl::adapt_airspeed_setpoint(const float control_interval,
 
-	if (!PX4_ISFINITE(_airspeed_slew_rate_controller.getState())) {
-		_airspeed_slew_rate_controller.setForcedValue(calibrated_airspeed_setpoint);
+	if (!PX4_ISFINITE(_airspeed_slew_rate_controller.getState())
+	    || _airspeed_slew_rate_controller.getState() < calibrated_min_airspeed || !_tecs_is_running) {
+		_airspeed_slew_rate_controller.setForcedValue(math::max(calibrated_min_airspeed, _airspeed_eas));
```
if _airspeed_slew_rate_controller.getState() < calibrated_min_airspeed you want to set it always to calibrated_min_airspeed not to _airspeed_eas. For the other two conditions, it is fine.

sfuhrer replied:
Good catch, done in

KonradRudin reviewed (commented):
Generally i like this cleanup, i think the parameters are better now.
About the IAS/CAS/EAS/TAS cleanup, i think we should do more (not necessarily here).

- Airspeed.msg: Why is an airspeed sensor already publishing the TAS? Then it would need the altitude and temperature to convert. That makes no sense to me. Also it is only used in the EKF, which could use the library function to convert from CAS to TAS there. 
  - I think the Airspeed.msg should only publish IAS and  maybe CAS, if the calibration data is available to the sensor.
  - Not sure what to do with calc_IAS_corrected(). Shouldn't that go into the CAS?

- AirspeedValidated.msg: It publishes the IAS, CAS and TAS and neglects EAS.
  - I get it that we assume that CAS equals EAS for low speed and altitude, but then we could rename it directly EAS and not CAS. Now we have several places in the code where modules assign _airspeed_eas = airspeed_validated.calibrated_airspeed_m_s which looks fishy. i think it would be better that airspeedValidated publishes all four airspeeds and internally make a function EAS_from_CAS() which is a one to one mapping. 
  - If you want, you could also only publish EAS and TAS since the others are nor used (and could be queried from the Airspeed message).

tstastny reviewed (changes requested):
Nice clean up :) Just a few comments

dagar commented:
So far looks good, I'm definitely in favor of getting rid of most of those circuit breakers and that FW_ARSP_MODE parameter wasn't great.

dagar commented on src/lib/systemlib/system_params.c:
```diff
@@ -216,6 +216,19 @@ PARAM_DEFINE_INT32(SYS_HAS_MAG, 1);
+ *
+ * @group System
+ * @min 0
+ * @max 1
```
Kind of, the calibration and workflow is missing/wrong for more than a single sensor.

sfuhrer replied:
Correct, we're not quite there atm.

dagar commented on src/modules/commander/HealthAndArmingChecks/checks/airspeedCheck.cpp:
```diff
@@ -37,24 +37,20 @@
 	airspeed_validated_s airspeed_validated;
 
-	if (_airspeed_validated_sub.copy(&airspeed_validated) && hrt_elapsed_time(&airspeed_validated.timestamp) < 1_s) {
+	if (_airspeed_validated_sub.copy(&airspeed_validated) && hrt_elapsed_time(&airspeed_validated.timestamp) < 2_s) {
```
Why is 2 seconds necessary?

sfuhrer replied:
>Not sure if out of scope here, but shouldn't we check here as well if all the declared airspeed sensor are present? And that the validated airspeed is coming from the set primary? So we can catch that all airspeed sensors are present and at least the primary is healthy (or better all are healthy)

Yes we should do that, the same way as it's done for other sensors. But right now we don't support more than 1 airspeed sensor so it's not needed yet.

sfuhrer replied:
>Why is 2 seconds necessary?

The airspeed selector only starts publishing 2s after boot: https://github.com/PX4/PX4-Autopilot/blob/8da106df6aae07bdd7fbf8675062a633b373dc8f/src/modules/airspeed_selector/airspeed_selector_main.cpp#L301. Maybe we also have a better fix for https://github.com/PX4/PX4-Autopilot/pull/14811 now?

sfuhrer commented on src/lib/parameters/param_translation.cpp:
```diff
@@ -42,7 +42,21 @@
+				node->i32 = 0;
+			}
+
+			strcpy(node->name, "FW_USE_AIRSPD");
```
Yes, good point. Done in 27d4407a

sfuhrer commented on src/modules/fw_pos_control/FixedwingPositionControl.cpp:
```diff
@@ -214,20 +205,23 @@ FixedwingPositionControl::airspeed_poll()
+			_airspeed_eas = airspeed_validated.calibrated_airspeed_m_s;
 
 			_eas2tas = constrain(airspeed_validated.true_airspeed_m_s / airspeed_validated.calibrated_airspeed_m_s, 0.9f, 2.0f);
+
+		} else {
+			airspeed_valid = false;
```
Before (on current main) it would stay valid forever if the received airspeed gets invalidated in air, set to NAN, but was finite before, no?

KonradRudin replied:
No, then https://github.com/PX4/PX4-Autopilot/blob/b13d9c929d7a4438de40ae05d72180dd533cb60d/src/modules/fw_pos_control/FixedwingPositionControl.cpp#L218 the timestamp is not set anymore and it should get invalidated afterwards here https://github.com/PX4/PX4-Autopilot/blob/b13d9c929d7a4438de40ae05d72180dd533cb60d/src/modules/fw_pos_control/FixedwingPositionControl.cpp#L228-L230 although that should not be in the else scope.

sfuhrer replied:
When you get an updated airspeed_validated input then it's validity should always overwrite _airspeed_valid immediately (to TRUE if finite, to FALSE if NAN).

sfuhrer commented on src/modules/airspeed_selector/airspeed_selector_main.cpp:
```diff
@@ -600,7 +600,7 @@ void AirspeedModule::select_airspeed_and_publish()
-	if (_param_airspeed_primary_index.get() > airspeed_index::DISABLED_INDEX &&
+	if (_param_airspeed_primary_index.get() >= airspeed_index::GROUND_MINUS_WIND_INDEX &&
 	    (_valid_airspeed_index < airspeed_index::FIRST_SENSOR_INDEX
 	     || _param_airspeed_primary_index.get() == airspeed_index::GROUND_MINUS_WIND_INDEX)) {
```
Yes, thanks! b2de2a29

sfuhrer commented on src/modules/commander/HealthAndArmingChecks/checks/airspeedCheck.cpp:
```diff
@@ -37,24 +37,20 @@
 void AirspeedChecks::checkAndReport(const Context &context, Report &reporter)
 {
-	if (circuit_breaker_enabled_by_val(_param_cbrk_airspd_chk.get(), CBRK_AIRSPD_CHK_KEY) ||
+	if (_param_sys_has_num_aspd.get() <= 0 ||
```
Well you could force safe it to -1, so better catch it here IMO.

sfuhrer commented on src/modules/commander/HealthAndArmingChecks/checks/airspeedCheck.cpp:
```diff
@@ -37,24 +37,20 @@
 	airspeed_validated_s airspeed_validated;
 
-	if (_airspeed_validated_sub.copy(&airspeed_validated) && hrt_elapsed_time(&airspeed_validated.timestamp) < 1_s) {
+	if (_airspeed_validated_sub.copy(&airspeed_validated) && hrt_elapsed_time(&airspeed_validated.timestamp) < 2_s) {
```
see https://github.com/PX4/PX4-Autopilot/pull/22510#discussion_r1431435899

sfuhrer commented on src/modules/fw_rate_control/FixedwingRateControl.hpp:
```diff
@@ -164,7 +164,7 @@ class FixedwingRateControl final : public ModuleBase<FixedwingRateControl>, publ
 		(ParamFloat<px4::params::FW_AIRSPD_STALL>) _param_fw_airspd_stall,
 		(ParamFloat<px4::params::FW_AIRSPD_TRIM>) _param_fw_airspd_trim,
-		(ParamInt<px4::params::FW_ARSP_MODE>) _param_fw_arsp_mode,
+		(ParamInt<px4::params::FW_USE_AIRSPD>) _param_fw_use_airspd,
```
Good catch, there were a couple more: 84025b017

sfuhrer commented:
@KonradRudin @tstastny addressed your comments.

@KonradRudin [your comment](https://github.com/PX4/PX4-Autopilot/pull/22510#pullrequestreview-1786840244) is noted and valid, we can do some followup clean ups for sure.

KonradRudin reviewed (approved):
Looks good to me.

sfuhrer commented:
Rebased on main and cleaned up the commits, done last SITL testing and param translation checks. LGTM.

sfuhrer commented:
@DonLakeFlyer thanks for raising it, I wasn't diligent enough, should check compatibility with QGC more often.
When ever possible I would though try to stay away from using PX4 parameters in QGC. Let's move the discussion for this specific issue here to https://github.com/mavlink/qgroundcontrol/issues/11252.
